### PR TITLE
NIAD-585: Referral Request missing elements

### PR DESCRIPTION
**Pull request by Broom94**

### CODE REVIEW

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/ProcedureRequestMapper.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+public class ProcedureRequestMapper {
+    public ProcedureRequest mapProcedureRequest(POCDMT000002UK01ClinicalDocument1 clinicalDocument) {
+        ProcedureRequest procedureRequest = new ProcedureRequest();
+        if (clinicalDocument.isSetComponentOf()) {
+            if (clinicalDocument.getComponentOf().getEncompassingEncounter() != null) {
+                if (clinicalDocument.getComponentOf().getEncompassingEncounter().isSetDischargeDispositionCode()) {
+                    if (clinicalDocument.getComponentOf().getEncompassingEncounter().getDischargeDispositionCode().isSetDisplayName()) {
+                        procedureRequest.setIdElement(newRandomUuid());
+                        String procedureReason =
+                            clinicalDocument.getComponentOf().getEncompassingEncounter().getDischargeDispositionCode().getDisplayName();
```
Comment (dawidskainos) [inline service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/ProcedureRequestMapper.java:24]: Please see the following comment:
"Disposition is located at /ClinicalDocument/componentOf/encompassingEncounter/discharge DispositionCode/@code – the code system is fixed to “2.16.840.1.113883.2.1.3.2. 4.17.325”
This should be mapped to ReferralRequest.supportingInfo(ProcedureRequest).code"

Let's map full CodeableConcept -  system, code, display

**Review by dawidskainos — CHANGES_REQUESTED** [review]

**Review by dawidskainos — DISMISSED** [review]

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/ProcedureRequestMapper.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+                    }
+                    if (clinicalDocument.getComponentOf().getEncompassingEncounter().getDischargeDispositionCode().isSetCode()) {
+                        coding.setCode(clinicalDocument.getComponentOf().getEncompassingEncounter().getDischargeDispositionCode()
+                            .getCode());
+                    }
+                    if (clinicalDocument.getComponentOf().getEncompassingEncounter().getDischargeDispositionCode().isSetCodeSystem()) {
+                        coding.setSystem(clinicalDocument.getComponentOf().getEncompassingEncounter().getDischargeDispositionCode()
+                            .getCodeSystem());
+                    }
+                    if (!StringUtils.isBlank(coding.getCode()) || !StringUtils.isBlank(coding.getDisplay())
```
Comment (bartek-sarul) [inline service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/ProcedureRequestMapper.java:36]: Use `StringUtils.isNotBlank(coding.getCode())` instead of negation
Also I think there was a helper `anyNotBlank` in `StringUtils` or in `CollectionUtils`

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/service/EncounterReportBundleService.java [hunk 3]
```diff
@@ -86,6 +99,25 @@ public Bundle createEncounterBundle(POCDMT000002UK01ClinicalDocument1 clinicalDo
...
     }
 
+    private Encounter addEncounterObjects(Encounter encounter, EncounterHelper encounterHelper,
+        POCDMT000002UK01ClinicalDocument1 clinicalDocument, List<QuestionnaireResponse> questionnaireResponseList) {
+        if (encounterHelper.getReferralRequest() != null) {
+            encounter.addIncomingReferral(new Reference(referralRequestMapper.addCondition(
+                encounterHelper.getReferralRequest(), clinicalDocument, encounter, questionnaireResponseList)));
+        }
+
+        if (encounterHelper.getAppointment() != null) {
```
Comment (bartek-sarul) [inline service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/service/EncounterReportBundleService.java:109]: If it's `getAppointment()` returns an `Optional` then we should assume it's invalid for it to be null.

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/service/EncounterReportBundleService.java [hunk 4]
```diff
@@ -86,6 +99,25 @@ public Bundle createEncounterBundle(POCDMT000002UK01ClinicalDocument1 clinicalDo
...
+            if (encounterHelper.getAppointment().isPresent()) {
```
Comment (bartek-sarul) [inline service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/service/EncounterReportBundleService.java:110]: If it's an `Optional` then do this
```
encounterHelper.getAppointment()
    .map(encounterAppointment -> appointmentMapper.addReferralRequest(encounterAppointment, encounterHelper.getReferralRequest()))
    .ifPresent(appointment -> {
        encounter.setAppointment(new Reference(appointment));
        encounter.setAppointmentTarget(appointment);
});
```

**Review by bartek-sarul — CHANGES_REQUESTED** [review]

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/ReferralRequestMapper.java [hunk 5]
```diff
@@ -44,12 +61,67 @@ public ReferralRequest mapReferralRequest(Encounter encounter, List<HealthcareSe
...
+            .setSubject(encounter.getSubject())
+            .setContext(new Reference(encounter))
+            .setEvidence(evidenceOf(questionnaireResponseList));
+        condition.setId(newRandomUuid());
+
+        return new Reference(condition);
+    }
+
+    private List<ConditionEvidenceComponent> evidenceOf(List<QuestionnaireResponse> questionnaireResponseList) {
+        List<ConditionEvidenceComponent> evidenceComponentList = new ArrayList<>();
```
Comment (bartek-sarul) [inline service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/ReferralRequestMapper.java:117]: [optional] Can be:
```
        return questionnaireResponseList.stream()
            .filter(Resource::hasId)
            .map(Resource::getId)
            .map(Reference::new)
            .map(reference -> new ConditionEvidenceComponent().addDetail(reference))
            .collect(Collectors.toList());
```

**Review by bartek-sarul — CHANGES_REQUESTED** [review]

**Review by bartek-sarul — APPROVED** [review]

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/util/CodeUtil.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+import org.hl7.fhir.dstu3.model.CodeableConcept;
+import org.hl7.fhir.dstu3.model.Coding;
+
+import lombok.experimental.UtilityClass;
+import uk.nhs.connect.iucds.cda.ucr.CE;
+
+@UtilityClass
+public class CodeUtil {
+
+    public CodeableConcept createCodeableConceptList(CE ce) {
```
Comment (dawidskainos) [inline service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/util/CodeUtil.java:12]: Shouldn't this be static if this is UtilityClass?

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/objects/EncounterHelper.java [hunk 7]
```diff
@@ -0,0 +1,18 @@
+package uk.nhs.adaptors.oneoneone.cda.report.objects;
```
Comment (dawidskainos) [inline service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/objects/EncounterHelper.java:1]: Let's not use such "helpers". I think it would be a better idea to extract necessary mappers calls out of EncounterMapper and only pass their outcome if necessary.

**Review by dawidskainos — CHANGES_REQUESTED** [review]

**Comment by dawidskainos** [comment]
I found two issues:
1) ReferralRequest.supportingInfo is not added to the bundle
2) ReferralRequest.reasonReference is not included in the bundle due to test data probably. Could you modify test data so that example reasonReference is in the outcome message?

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/EncounterMapper.java [hunk 8]
```diff
@@ -151,26 +142,10 @@ private void setSubject(Encounter encounter, POCDMT000002UK01ClinicalDocument1 c
...
-        if (appointment.isPresent()) {
-            encounter.setAppointment(new Reference(appointment.get()));
-            encounter.setAppointmentTarget(appointment.get());
-        }
-    }
-
-    private void setEpisodeOfCare(Encounter encounter, POCDMT000002UK01ClinicalDocument1 clinicalDocument) {
-        Optional<EpisodeOfCare> episodeOfCare = episodeOfCareMapper.mapEpisodeOfCare(clinicalDocument, encounter.getSubject());
-        episodeOfCare.ifPresent(ofCare -> encounter.addEpisodeOfCare(new Reference(ofCare)));
+        appointmentService.retrieveAppointment(patient, clinicalDocument).ifPresent(appointment ->
```
Comment (bartek-sarul) [inline service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/EncounterMapper.java:147]: Could be a bit better with 
```
.map(Reference::new)
.ifPresent(encounter::setAppointment)
```

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/ConditionMapper.java [hunk 9]
```diff
@@ -61,21 +76,7 @@ public Condition mapCondition(POCDMT000002UK01ClinicalDocument1 clinicalDocument
...
-                                        coding.setCode(language.getCode());
-                                    }
-                                    if (!StringUtils.isBlank(language.getSystem())) {
-                                        coding.setSystem(language.getSystem());
-                                    }
-                                    if (coding.hasCode() || coding.hasSystem() || coding.hasDisplay()) {
-                                        condition.setCode(new CodeableConcept(coding));
-                                    }
-                                });
+                                condition.setCode(new CodeableConcept().setText(component.getSection().getLanguageCode().getCode()));
```
Comment (dawidskainos) [inline service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/ConditionMapper.java:79]: What is the reason behind this change? It looks like the entire CodeableConcept was populated and now only text is (which was already reported as an issue in other cases)

**Review by dawidskainos — CHANGES_REQUESTED** [review]

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/ConditionMapper.java (hunk 9 above)
Comment (Broom94) [reply service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/ConditionMapper.java:79]: We previously discussed this to remove language enum unless I have misunderstood?

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/ProcedureRequestMapper.java [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+@RequiredArgsConstructor
+public class ProcedureRequestMapper {
+    public ProcedureRequest mapProcedureRequest(POCDMT000002UK01ClinicalDocument1 clinicalDocument, Reference patient,
+        ReferralRequest referralRequest) {
+        ProcedureRequest procedureRequest = new ProcedureRequest();
+        if (clinicalDocument.isSetComponentOf()) {
+            if (clinicalDocument.getComponentOf().getEncompassingEncounter() != null) {
+                if (clinicalDocument.getComponentOf().getEncompassingEncounter().isSetDischargeDispositionCode()) {
+                    Coding coding = new Coding();
+                    CE dichargeCode = clinicalDocument.getComponentOf().getEncompassingEncounter().getDischargeDispositionCode();
```
Comment (saddiqs1) [inline service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/ProcedureRequestMapper.java:27]: Potential typo on `dichargeCode`? Should it be `dischargeCode`?

Comment (dawidskainos) [reply service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/ProcedureRequestMapper.java:27]: good catch

**Review by dawidskainos — DISMISSED** [review]

**Review by dawidskainos — APPROVED** [review]

**Review by saddiqs1 — APPROVED** [review]
